File: java_kotlin_playground/src/main/java/leetcode/_200.java
package leetcode;

public class _200 {
	private int row;
	private int col;

	public int solution(char[][] grid) {
		int answer = 0;
		this.row = grid.length;
		this.col = grid[0].length;

		for (int i = 0; i < row; i++) {
			for (int j = 0; j < col; j++) {
				if (grid[i][j] == '1') {
					answer += 1;
					moveTo(grid, i, j);
				}
			}
		}

		return answer;
	}

	private void moveTo(char[][] grid, int row, int col) {
		if (row == this.row || row == -1 ||
			col == this.col || col == -1 ||
			grid[row][col] != '1') {
			return;
		}

		grid[row][col] = 'X';

		moveTo(grid, row + 1, col);
		moveTo(grid, row, col + 1);
		moveTo(grid, row - 1, col);
		moveTo(grid, row, col - 1);
	}
}

File: java_kotlin_playground/src/test/java/leetcode/_200.kt
package leetcode

class _200 {
    private var row = 0;
    private var col = 0;

    fun numIsLands(grid: Array<CharArray>): Int {
        val answer = 0
        row = grid.size
        col = grid[0].size

        grid.forEachIndexed { y, row ->
            row.forEachIndexed { x, value ->
                if (value == '1') moveTo(grid, y, x)
            }
        }

        return answer
    }

    private fun moveTo(grid: Array<CharArray>, y: Int, x: Int) {
        if (row == this.row || row == -1 ||
            col == this.col || col == -1 ||
            grid[row][col] != '1'
        ) {
            return
        }

        grid[row][col] = 'X'

        moveTo(grid, row + 1, col)
        moveTo(grid, row, col + 1)
        moveTo(grid, row - 1, col)
        moveTo(grid, row, col - 1)
    }
}